Mathlib/Algebra/Polynomial/Expand.lean

```diff
@@ -176,6 +176,12 @@ theorem expand_aeval {A : Type*} [Semiring A] [Algebra R A] (p : ℕ) (P : R[X])
   refine Polynomial.induction_on P (fun a => by simp) (fun f g hf hg => ?_) fun n a _ => by simp
   rw [map_add, aeval_add, aeval_add, hf, hg]
 
+@[gcongr]
+theorem expand_dvd {a b : R[X]} (p : ℕ) (h : a ∣ b) : expand R p a ∣ expand R p b := by
```

Comment:
Empirically, explicit arguments are only really useful for rewriting lemmas, which this one is not
```suggestion
theorem expand_dvd {a b : R[X]} {p : ℕ} (h : a ∣ b) : expand R p a ∣ expand R p b := by
```

Reply:
I would expect the explicit argument to be useful for `have := expand_dvd p h` or `obtain`. On the other hand, for rewriting lemmas I'd usually keep arguments appearing on both sides implicit, since they are not needed most of the time when using `rw` and making them implicit saves some underscores (or even parentheses) when you occasionally use the lemmas in term mode. Can you explain your rationale (or post a link to your explanation if you've done it before)?

Reply:
It's possibly an artifact of my style but here are some arguments:
1. Non-rewriting lemmas are almost always used in a context where there's enough information for Lean to infer them by unification.
2. In the rare cases where there isn't enough info, it usually means you should add a type ascription for your reader to understand, which is better than filling in an argument of unclear purpose. Eg I would consider `obtain ⟨q, hq⟩ : expand R p a ∣ expand R p b := expand_dvd h` better style than `obtain ⟨q, hq⟩ := expand_dvd p h`.
3. Non-rewriting lemmas are very likely to be used in proof terms, where extra underscores and brackets are painful.
4. Rewriting lemmas are almost always used with `rw`, where extra explicit arguments don't hurt when unnecessary and come in handy when needed to disambiguate between several occurrences.

Reply:
Thanks for the detailed explanations!
I agree that variables appearing in the conclusion of non-rewriting lemmas can be made implicit. Explicit arguments in rewriting lemmas for disambiguation is also a good point. Regarding point 5, what do you think about the many uses of an iff lemma concatenated with `.trans` in [these search results](https://github.com/search?q=repo%3Aleanprover-community%2Fmathlib4%20iff.trans&type=code)? I don't think these should be made calc blocks.

For this particular lemma, I realize we don't need it as it's a special case of `map_dvd`.